agent/requestor-sdk/src/requestor.rs

```diff
@@ -92,365 +130,481 @@ impl Requestor {
+        log::info!("releasing allocation");
+        if let Err(e) = payment_manager.send(ReleaseAllocation).await {
+            log::warn!("unable to release allocation: {:?}", e);
+        }
 
         Ok(())
     }
 
     async fn create_demand(&self) -> Result<Demand> {
-        // "golem.node.debug.subnet" == "mysubnet", TODO
         let (digest, url) = self.task_package.publish().await?;
         let url_with_hash = format!("hash:sha3:{}:{}", digest, url);
         let constraints = self.constraints.clone().and(constraints![
             "golem.runtime.name" == self.image_type.runtime_name(),
             "golem.runtime.version" == self.image_type.runtime_version().to_string(),
+            "golem.node.debug.subnet" == self.subnet.clone(),
         ]);
 
         log::debug!("srv.comp.task_package: {}", url_with_hash);
 
+        let deadline = chrono::Utc::now() + chrono::Duration::from_std(self.timeout.clone())?;
         let demand = Demand::new(
             serde_json::json!({
-                "golem": {
-                    "node.id.name": self.name,
-                    "srv.comp.task_package": url_with_hash,
-                    "srv.comp.expiration":
-                        (chrono::Utc::now() + chrono::Duration::minutes(10)).timestamp_millis(), // TODO
-                },
+                "golem.node.id.name": self.name,
+                "golem.node.debug.subnet": self.subnet.clone(),
+                "golem.srv.comp.task_package": url_with_hash,
+                "golem.srv.comp.expiration": deadline.timestamp_millis(),
             }),
             constraints.to_string(),
         );
 
         Ok(demand)
     }
+}
 
-    async fn create_agreement(
-        market_api: MarketRequestorApi,
-        activity_api: ActivityRequestorApi,
-        payment_manager: Addr<PaymentManager>,
-        proposal: Proposal,
-        exe_script: ExeScript,
-    ) -> Result<(String, String)> {
-        let id = proposal.proposal_id()?;
-        let issuer = proposal.issuer_id()?;
-        log::debug!("hello issuer: {}", issuer);
-
-        let agreement = AgreementProposal::new(
-            id.clone(),
-            chrono::Utc::now() + chrono::Duration::minutes(10), /* TODO */
-        );
-
-        log::info!("creating agreement");
-
-        /* TODO handle errors */
-        let r = market_api.create_agreement(&agreement).await;
-
-        log::info!("create agreement result: {:?}; confirming", r);
-
-        let _ = market_api.confirm_agreement(&id).await;
-
-        log::info!("waiting for approval");
-
-        let _ = market_api.wait_for_approval(&id, Some(10.0)).await;
-
-        log::info!("approval received for agreement: {}", id);
+async fn process_market_events(
+    requestor: Addr<Requestor>,
+    market_api: MarketRequestorApi,
+    subscription_id: String,
+    demand: Demand,
+    mut tx: mpsc::Sender<Proposal>,
+) {
+    log::info!("processing market events");
+    'outer: loop {
```

Comment:
Design question: if we only ever trigger a combination of negative paths here, is it possible we've just created an infinite loop that never breaks?

Reply:
`Ok(ComputationState::Finished) => break 'outer,`

`Finished` means just "finished", whatever the reason (successful, timeout, etc.)